src/bluez/peripheral.rs

```diff
@@ -0,0 +1,245 @@
+        let characteristic_info = self.characteristic_info(characteristic)?;
+        Ok(self
+            .session
+            .read_characteristic_value(&characteristic_info.id)
+            .await?)
+    }
+
+    async fn subscribe(&self, characteristic: &Characteristic) -> Result<()> {
+        let characteristic_info = self.characteristic_info(characteristic)?;
+        Ok(self.session.start_notify(&characteristic_info.id).await?)
+    }
+
+    async fn unsubscribe(&self, characteristic: &Characteristic) -> Result<()> {
+        let characteristic_info = self.characteristic_info(characteristic)?;
+        Ok(self.session.stop_notify(&characteristic_info.id).await?)
+    }
+
+    async fn notifications(&self) -> Result<Pin<Box<dyn Stream<Item = ValueNotification>>>> {
+        let device_id = self.device.clone();
+        let events = self.session.device_event_stream(&device_id).await?;
+        let characteristics = self.characteristics.clone();
+        Ok(Box::pin(events.filter_map(move |event| {
+            ready(value_notification(
+                event,
+                &device_id,
+                characteristics.clone(),
+            ))
+        })))
+    }
+}
+
+fn value_notification(
+    event: BluetoothEvent,
+    device_id: &DeviceId,
+    characteristics: Arc<Mutex<Vec<CharacteristicInfo>>>,
+) -> Option<ValueNotification> {
+    match event {
+        BluetoothEvent::Characteristic {
+            id,
+            event: CharacteristicEvent::Value { value },
+        } if id.service().device() == *device_id => {
+            let characteristics = characteristics.lock().unwrap();
+            let uuid = characteristics
+                .iter()
+                .find(|characteristic| characteristic.id == id)?
+                .uuid
+                .into();
+            Some(ValueNotification {
+                uuid,
+                handle: None,
+                value,
+            })
+        }
+        _ => None,
+    }
+}
+
+impl From<WriteType> for bluez_async::WriteType {
+    fn from(write_type: WriteType) -> Self {
+        match write_type {
+            WriteType::WithoutResponse => bluez_async::WriteType::WithoutResponse,
+            WriteType::WithResponse => bluez_async::WriteType::WithResponse,
+        }
+    }
+}
+
+impl From<&MacAddress> for BDAddr {
+    fn from(mac_address: &MacAddress) -> Self {
+        mac_address.to_string().parse().unwrap()
+    }
+}
+
+impl From<bluez_async::AddressType> for AddressType {
+    fn from(address_type: bluez_async::AddressType) -> Self {
+        match address_type {
+            bluez_async::AddressType::Public => AddressType::Public,
+            bluez_async::AddressType::Random => AddressType::Random,
+        }
+    }
+}
+
+impl From<&CharacteristicInfo> for Characteristic {
+    fn from(characteristic: &CharacteristicInfo) -> Self {
+        Characteristic {
+            start_handle: 0,
+            end_handle: 0,
+            value_handle: 0,
```

Comment:
If we're not using these properties anywhere and the API is getting re-written. I think it's safe to drop these from `PeripheralProperties`

Reply:
Good point, done.